### DISCUSSION TITLE
How should you structure code to avoid import cycles?

### DISCUSSION THREAD

**Discussion by parched** [microsoft/pyright#4425] [Q&A] [answered]
I have some code that looks something like this:
```
# my_mod/__init__.py

from _a import entry_point as entry_point
```

```
# my_mod/_a.py

from . import _b

def entry_point() -> str:
    return _b.B
```

```
# my_mod/_b.py

B = "B"
```
There's an import cycle between `__init__.py` and `_a.py`.

The actual files are much larger, so I want to keep them broken down. That's why `entry_point` isn't defined in `__init__.py` as perceived by the user of this module. Can I achieve this another way without an import cycle?

**Comment by erictraut** [comment, marked as answer]
In `_a.py`, import the symbol `B` directly from `_b` rather than importing it indirectly through `__init__.py`.

```python
from _b import B

def entry_point() -> str:
    return B
```

or

```python
from ._b import B

def entry_point() -> str:
    return B
```

**Reply by parched** [reply]
In this simple case it works, but I have lots of things in `_b` and the names of them only make sense in the context of `_a` with the `_b` prefix.

E.g., it's more like this
```python
from . import _b, _c

something = _b.x + _c.x, _b.y + _c.y, _b.z + _c.z
```

I could do
```python
from ._b import x as b_x, y as b_y, z as b_z
from ._c import x as c_x, y as c_y, z as c_z

something = b_x + c_x, b_y + c_y, b_z + c_z
```
but it's more tedious and less autocomplete friendly.

Could pyright check that `_b` is actually a submodule and not some name defined in `__init__.py` and not flag it as cyclic if that's true.
As far as I'm aware the `from ._b import B` is still technically an import cycle, but a benign one like my original code.

**Reply by erictraut** [reply]
The `reportImportCycle` check is opt-in and is not enabled by default even in strict mode. It's very opinionated in how it works. There are reasonable arguments to be made that "cycle" should be defined differently than the current implementation, but opinions will differ here.

I added this check specifically for my team's code base because we were struggling to enforce architectural layering. Import cycles are admittedly not something that a type checker should be checking for, but I wasn't able to find a better solution at the time. Perhaps there are better options available now.

If you don't like the way the import cycle check works in its current implementation, I recommend disabling it in pyright and looking for a linter option that better suits your needs.
